// light/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useLogger = exports.withConnect = exports.createRoute = exports.createServer = exports.createRouteTest = exports.createTest = exports.createMiddlewareConfig = exports.createLoggerConfig = exports.createConfig = void 0;
var config_1 = require("@lightjs/config");
Object.defineProperty(exports, "createConfig", { enumerable: true, get: function () { return config_1.createConfig; } });
Object.defineProperty(exports, "createLoggerConfig", { enumerable: true, get: function () { return config_1.createLoggerConfig; } });
Object.defineProperty(exports, "createMiddlewareConfig", { enumerable: true, get: function () { return config_1.createMiddlewareConfig; } });
var test_1 = require("@lightjs/test");
Object.defineProperty(exports, "createTest", { enumerable: true, get: function () { return test_1.createTest; } });
Object.defineProperty(exports, "createRouteTest", { enumerable: true, get: function () { return test_1.createRouteTest; } });
var server_1 = require("@lightjs/server");
Object.defineProperty(exports, "createServer", { enumerable: true, get: function () { return server_1.createServer; } });
var core_1 = require("@lightjs/core");
Object.defineProperty(exports, "createRoute", { enumerable: true, get: function () { return core_1.createRoute; } });
Object.defineProperty(exports, "withConnect", { enumerable: true, get: function () { return core_1.withConnect; } });
var logger_1 = require("@lightjs/logger");
Object.defineProperty(exports, "useLogger", { enumerable: true, get: function () { return logger_1.useLogger; } });
