// @lightjs/core/lib/createRoute/isServerless.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isServerless = exports.isNextJS = exports.isVercel = exports.isNow = exports.isRunKit = exports.isAWS = exports.isNetlify = void 0;
const { LIGHT_ENV } = process.env;
exports.isNetlify = (LIGHT_ENV === null || LIGHT_ENV === void 0 ? void 0 : LIGHT_ENV.toLowerCase()) === 'netlify';
exports.isAWS = (LIGHT_ENV === null || LIGHT_ENV === void 0 ? void 0 : LIGHT_ENV.toLowerCase()) === 'aws';
exports.isRunKit = (LIGHT_ENV === null || LIGHT_ENV === void 0 ? void 0 : LIGHT_ENV.toLowerCase()) === 'runkit';
exports.isNow = (LIGHT_ENV === null || LIGHT_ENV === void 0 ? void 0 : LIGHT_ENV.toLowerCase()) === 'now';
exports.isVercel = (LIGHT_ENV === null || LIGHT_ENV === void 0 ? void 0 : LIGHT_ENV.toLowerCase()) === 'vercel';
exports.isNextJS = (LIGHT_ENV === null || LIGHT_ENV === void 0 ? void 0 : LIGHT_ENV.toLowerCase()) === 'nextjs';
exports.isServerless = exports.isNetlify || exports.isAWS || exports.isRunKit || exports.isNow || exports.isVercel || exports.isNextJS;
