// @lightjs/cli/lib/commands/start.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.builder = exports.desc = exports.command = void 0;
const emojic_1 = __importDefault(require("emojic"));
const chalk_1 = __importDefault(require("chalk"));
const config_1 = require("@lightjs/config");
const server_1 = require("@lightjs/server");
const logger_1 = require("@lightjs/logger");
exports.command = 'start';
exports.desc = 'start a production server';
exports.builder = {
    port: {
        alias: 'p',
        description: 'specify which port the server should run on',
    },
};
const handle = async (argv) => {
    const ts = (0, config_1.isTypescript)();
    if (ts) {
        require('ts-node').register(); // eslint-disable-line
    }
    const logger = (0, logger_1.useFrameworkLogger)();
    const middlewareConfig = (0, config_1.importMiddlewareConfig)();
    const globalMiddleware = (middlewareConfig === null || middlewareConfig === void 0 ? void 0 : middlewareConfig().global) || [];
    logger.info(`[ ${chalk_1.default.redBright('start')} ] ${emojic_1.default.fire} igniting the server ${emojic_1.default.fire}`);
    const app = (0, server_1.createServer)({
        middleware: globalMiddleware,
    });
    const { HOST = '0.0.0.0' } = process.env;
    let { PORT = 3000 } = process.env;
    if (argv.port) {
        PORT = argv.port;
    }
    app.server.listen(PORT, HOST, () => {
        logger.info(`[ ${chalk_1.default.magentaBright('listening')} ] on port ${PORT}`);
    });
};
const handler = (argv) => {
    handle(argv).catch((err) => {
        console.error(err);
        process.exit(1);
    });
};
exports.handler = handler;
